Implement Observable#first

maraisr opened the pull request:
#### f3f3afde18272fdabcd5901b755b47bef3b77da2
<pre>
Implement Observable#first
<a href="https://bugs.webkit.org/show_bug.cgi?id=277509">https://bugs.webkit.org/show_bug.cgi?id=277509</a>

Reviewed by Chris Dumez and Darin Adler.

This change introduces the `.first` promise-returning operator
for Observables, using the InternalObserver and subscribing immediately.
This is required by the specification <a href="https://wicg.github.io/observable/#dom-observable-first">https://wicg.github.io/observable/#dom-observable-first</a>

Because promise-returning operators subscribe immediatly in the runtime,
as opposed to user-land. We made a change to the Subscriber class to
now hold a reference to it&apos;s options, specifically any AbortSignals.

* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-last.any.js: Rebaselined.
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-first.any-expected.txt: Rebaselined.
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-first.any.worker-expected.txt: Rebaselined.
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-first.any.js:
  AbortErrors&apos; description text is not normative,
   as such do not assert against it.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Observable.h:
* Source/WebCore/dom/Observable.cpp:
(WebCore::Observable::first): Added.
(WebCore::Observable::subscribeInternal):
  Move the abort signal following logic inside the Subscriber class.
* Source/WebCore/dom/Observable.idl:
  Exposes `Promise&lt;any&gt; first(options)` as a promise-returning operator.
* Source/WebCore/dom/Subscriber.h:
* Source/WebCore/dom/Subscriber.cpp:
(WebCore::Subscriber::create):
(WebCore::Subscriber::Subscriber):
  When a subscriber is created, it should own the options for the
  lifetime of the subscriber.
* Source/WebCore/dom/InternalObserverLast.cpp:
  Addressed diff feedback for `.first` in here as well.
* Source/WebCore/dom/InternalObserverLast.h:
* Source/WebCore/dom/InternalObserverFirst.cpp: Added.
(WebCore::createInternalObserverOperatorFirst):
* Source/WebCore/dom/InternalObserverFirst.h: Added.

Canonical link: <a href="https://commits.webkit.org/285650@main">https://commits.webkit.org/285650@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e8261a85c17bcd5fbbc3f55575b19766f869db7e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73383 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52812 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26191 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77629 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24621 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/75498 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/61971 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/595 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57644 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16103 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76450 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/61971 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63133 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38062 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/61971 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20615 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22950 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/61971 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20968 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79266 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/698 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/201 "Found 1 new test failure: http/tests/paymentrequest/payment-response-payerPhone-attribute.https.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66064 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/840 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63145 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65356 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16157 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9185 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7366 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/662 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/3399 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/692 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/676 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/694 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

maraisr commented on Source/WebCore/dom/Observable.h:
```diff
@@ -62,6 +62,10 @@ class Observable final : public ScriptWrappable, public RefCounted<Observable> {
 
     Ref<Observable> drop(ScriptExecutionContext&, uint64_t);
 
+    // Promise-returning operators.
```
Let's see, there may be conflicts with https://github.com/WebKit/WebKit/pull/32099. But should be okay, Ill resolve resolve with whatever lands first.

maraisr commented on Source/WebCore/dom/Observable.idl:
```diff
@@ -40,4 +41,8 @@ interface Observable {
   [CallWith=CurrentScriptExecutionContext] Observable take(unsigned long long amount);
 
   [CallWith=CurrentScriptExecutionContext] Observable drop(unsigned long long amount);
+
```
Let's see, there may be conflicts with https://github.com/WebKit/WebKit/pull/32099. But should be okay, Ill resolve resolve with whatever lands first.

darinadler commented on Source/WebCore/dom/InternalObserverFirst.cpp:
```diff
@@ -0,0 +1,111 @@
+{
+    Ref signal = AbortSignal::create(&context);
+
+    Vector<Ref<AbortSignal>> dependentSignals = { signal };
```
`WTFMove(signal)` or do away with the local variable entirely and do it on one line

maraisr replied:
Actually a good catch here mate. Thank you. 🙏🏻 Pointed that there was in fact a bug, I will need your help with. Spun far too long on this, and cannot figure it out.... 😞 

We really want to pass this `signal` down into `InternalObserverFirst::create`, as opposed too the `dependentSignal`. **BUT!** When I do that, [this promise hangs](https://github.com/WebKit/WebKit/blob/2d7df09cf3f08a332003e47dd0664f4c3423604a/LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-first.any.js#L66), and wpt times out. The abort, never rejects the promise, because:

Through a lot of debugging, our `dependentSignal->addAlgorithm` never fires. I did verify that the [`m_dependentSignals`](https://github.com/WebKit/WebKit/blob/2d7df09cf3f08a332003e47dd0664f4c3423604a/Source/WebCore/dom/AbortSignal.cpp#L122)'s `computedSize()` does increment for this test specifically. But then right before [this line](https://github.com/WebKit/WebKit/blob/2d7df09cf3f08a332003e47dd0664f4c3423604a/Source/WebCore/dom/AbortSignal.cpp#L137), that count is zero.

I suspect its lifetimes? Do I need to WTFMove something?

Note; the current HEAD diff here has that `Ref signal = AbortSignal::create(&context);` removed, as per your suggestion to inline it. But think that is wrong [as per the spec, point 3](https://wicg.github.io/observable/#dom-observable-first). 

Though the tests pass! 🎉 and if we're happy with that, we can move past this minor detail.

eg:

```diff
Ref signal = AbortSignal::create(&context);
Vector<Ref<AbortSignal>> dependentSignals = { signal };
if (UNLIKELY(options.signal))
    dependentSignals.append(*options.signal);
auto dependentSignal = AbortSignal::any(context, dependentSignals);

if (UNLIKELY(dependentSignal->aborted()))
    return promise->reject<IDLAny>(dependentSignal->reason().getValue());

dependentSignal->addAlgorithm([promise](JSC::JSValue reason) {
    promise->reject<IDLAny>(reason);
});

- auto observer = InternalObserverFirst::create(context, dependentSignal, WTFMove(promise));
+ auto observer = InternalObserverFirst::create(context, signal, WTFMove(promise)); // 💥, now dependentSignal->addAlgorithm never fires.

observable->subscribeInternal(context, observer, SubscribeOptions { .signal = dependentSignal.ptr() });
```

rreno replied:
I read Darin's suggestion as pick one of these two ways to populate the Vector:

```C++
Ref signal = AbortSignal::create(&context);
Vector<Ref<AbortSignal>> dependentSignals = { WTFMove(signal) };
```

or 

```C++
Vector<Ref<AbortSignal>> dependentSignals = { AbortSignal::create(&context) };
```

To be honest I don't know enough about this API to know why your current branch has a bug here but you did change the composition of `dependentSignals` by initializing it as an empty array now.

darinadler replied:
Yes @rreno understood me correctly. I wasn't offering deeper insight about what is correct or not with the code.

darinadler commented on Source/WebCore/dom/InternalObserverFirst.cpp:
```diff
@@ -0,0 +1,111 @@
+
+    auto observer = InternalObserverFirst::create(context, dependentSignal, WTFMove(promise));
+
+    observable->subscribeInternal(context, observer, SubscribeOptions { .signal = &dependentSignal.get() });
```
`dependentSignal.ptr()`

maraisr commented on Source/WebCore/dom/InternalObserverFirst.cpp:
```diff
@@ -0,0 +1,113 @@
+
+    // FIXME: We really should be passing `signal` down instead of `dependentSignal`. As per point 3 here https://wicg.github.io/observable/#dom-observable-first
+    //    but we have this failing test as a result: https://github.com/WebKit/WebKit/blob/2d7df09cf3f08a332003e47dd0664f4c3423604a/LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-first.any.js#L66
+    auto observer = InternalObserverFirst::create(context, dependentSignal, WTFMove(promise));
```
@darinadler @cdumez Id like to focus your attention this line specifically in relation to this previous comment: https://github.com/WebKit/WebKit/pull/35261#discussion_r1804687496

cdumez replied:
Yes, for what it's worth, I agree that we should be passing `signal` here, not `dependentSignal`, per the spec.

maraisr replied:
@cdumez I have change this line to read _as we'd expect_.

maraisr commented on Source/WebCore/dom/InternalObserverFirst.cpp:
```diff
@@ -0,0 +1,113 @@
+void createInternalObserverOperatorFirst(ScriptExecutionContext& context, Ref<Observable> observable, SubscribeOptions options, Ref<DeferredPromise>&& promise)
+{
+    Ref signal = AbortSignal::create(&context);
+    Vector<Ref<AbortSignal>> dependentSignals = { WTFMove(signal) };
```
@darinadler I applied your suggestion for the inline signal here. Though left it as 2 lines, to reenforce my `FIXME` comment below.

Do note, that when I pass the `signal` as suggested below in the `FIXME`, together with the `WTFMove`, I do get a runtime crash like:

```
ASSERTION FAILED: m_ptr
12:19:36.125 3412   WebKit/WebKitBuild/Debug/usr/local/include/wtf/Ref.h(135) : T &WTF::Ref<WebCore::AbortSignal>::get() const [T = WebCore::AbortSignal, _PtrTraits = WTF::RawPtrTraits<WebCore::AbortSignal>, RefDerefTraits = WTF::DefaultRefDerefTraits<WebCore::AbortSignal>]
```

darinadler replied:
Sounds like you did something different than what I meant. What I meant was:
```
Vector<Ref<AbortSignal>> dependentSignals { AbortSignal::create(&context) };
```
Or this simpler form might even compile:
```
Vector dependentSignals { AbortSignal::create(&context) };
```

maraisr replied:
Yeah. That would work, given how [I am not passing the right signal](https://github.com/WebKit/WebKit/pull/35261#discussion_r1809770841). But then, whats the point in creating an inline signal in the first place if its not used at all.

But when passing the right signal down, and your suggestion — I do get that crash.

darinadler replied:
The assertion failure you mention above is what you'd get if you use `signal` after moving from it. You can only move a value once and must not use it after moving. I don't know exactly where the error is but I'm sure we can find it.

cdumez commented on Source/WebCore/dom/InternalObserverFirst.cpp:
```diff
@@ -0,0 +1,113 @@
+private:
+    void next(JSC::JSValue value) final
+    {
+        (Ref { m_promise })->resolve<IDLAny>(value);
```
You don't need the parentheses around `Ref { m_promise }`

cdumez commented on Source/WebCore/dom/InternalObserverFirst.cpp:
```diff
@@ -0,0 +1,113 @@
+    void next(JSC::JSValue value) final
+    {
+        (Ref { m_promise })->resolve<IDLAny>(value);
+        (Ref { m_signal })->signalAbort(JSC::jsUndefined());
```
ditto.

cdumez commented on Source/WebCore/dom/InternalObserverFirst.cpp:
```diff
@@ -0,0 +1,113 @@
+
+    void error(JSC::JSValue value) final
+    {
+        (Ref { m_promise })->reject<IDLAny>(value);
```
ditto.

cdumez commented on Source/WebCore/dom/InternalObserverFirst.cpp:
```diff
@@ -0,0 +1,113 @@
+    void complete() final
+    {
+        InternalObserver::complete();
+        (Ref { m_promise })->reject(Exception { ExceptionCode::RangeError, "No values in Observable"_s });
```
ditto.

cdumez commented on Source/WebCore/dom/Observable.cpp:
```diff
@@ -120,6 +121,11 @@ Ref<Observable> Observable::drop(ScriptExecutionContext& context, uint64_t amoun
     return create(createSubscriberCallbackDrop(context, *this, amount));
 }
 
+void Observable::first(ScriptExecutionContext& context, SubscribeOptions options, Ref<DeferredPromise>&& promise)
```
We should be passing a `const SubscribeOptions&` or a `SubscribeOptions&&`, avoid passing by value.

cdumez commented on Source/WebCore/dom/Observable.cpp:
```diff
@@ -120,6 +121,11 @@ Ref<Observable> Observable::drop(ScriptExecutionContext& context, uint64_t amoun
+    return createInternalObserverOperatorFirst(context, *this, options, WTFMove(promise));
+}
+
 void Observable::last(ScriptExecutionContext& context, SubscribeOptions options, Ref<DeferredPromise>&& promise)
```
ditto.

cdumez commented on Source/WebCore/dom/InternalObserverFirst.cpp:
```diff
@@ -0,0 +1,113 @@
+    Ref<DeferredPromise> m_promise;
+};
+
+void createInternalObserverOperatorFirst(ScriptExecutionContext& context, Ref<Observable> observable, SubscribeOptions options, Ref<DeferredPromise>&& promise)
```
`Ref<Observable>` -> `Observable&`. No need to cause ref-counting churn here.
`SubscribeOptions` -> `const SubscribeOptions&`

cdumez commented on Source/WebCore/dom/InternalObserverFirst.cpp:
```diff
@@ -0,0 +1,113 @@
+    Ref signal = AbortSignal::create(&context);
+    Vector<Ref<AbortSignal>> dependentSignals = { WTFMove(signal) };
+
+    if (UNLIKELY(options.signal))
```
I don't think we need UNLIKELY

cdumez commented on Source/WebCore/dom/InternalObserverFirst.cpp:
```diff
@@ -0,0 +1,113 @@
+        dependentSignals.append(*options.signal);
+    auto dependentSignal = AbortSignal::any(context, dependentSignals);
+
+    if (UNLIKELY(dependentSignal->aborted()))
```
ditto.

cdumez commented on Source/WebCore/dom/InternalObserverFirst.cpp:
```diff
@@ -0,0 +1,113 @@
+
+    if (UNLIKELY(options.signal))
+        dependentSignals.append(*options.signal);
+    auto dependentSignal = AbortSignal::any(context, dependentSignals);
```
`auto` -> `Ref`

cdumez replied:
`auto` -> `Ref`

cdumez commented on Source/WebCore/dom/InternalObserverFirst.cpp:
```diff
@@ -0,0 +1,113 @@
+
+class InternalObserverFirst final : public InternalObserver {
+public:
+    static Ref<InternalObserverFirst> create(ScriptExecutionContext& context, Ref<AbortSignal> signal, Ref<DeferredPromise>&& promise)
```
`Ref<AbortSignal>&&`

cdumez commented on Source/WebCore/dom/InternalObserverFirst.cpp:
```diff
@@ -0,0 +1,113 @@
+public:
+    static Ref<InternalObserverFirst> create(ScriptExecutionContext& context, Ref<AbortSignal> signal, Ref<DeferredPromise>&& promise)
+    {
+        Ref internalObserver = adoptRef(*new InternalObserverFirst(context, signal, WTFMove(promise)));
```
`WTFMove(signal)`

cdumez commented on Source/WebCore/dom/InternalObserverFirst.cpp:
```diff
@@ -0,0 +1,113 @@
+    {
+    }
+
+    InternalObserverFirst(ScriptExecutionContext& context, Ref<AbortSignal> signal, Ref<DeferredPromise>&& promise)
```
`Ref<AbortSignal>&&`

cdumez commented on Source/WebCore/dom/InternalObserverFirst.cpp:
```diff
@@ -0,0 +1,113 @@
+
+    InternalObserverFirst(ScriptExecutionContext& context, Ref<AbortSignal> signal, Ref<DeferredPromise>&& promise)
+        : InternalObserver(context)
+        , m_signal(signal)
```
`, m_signal(WTFMove(signal))`

maraisr commented on LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-first.any-expected.txt:
```diff
@@ -1,7 +1,9 @@
+PASS first(): Promise resolves with the first value from the source Observable
+PASS first(): Promise rejects with the error emitted from the source Observable
+PASS first(): Promise rejects with RangeError when source Observable completes without emitting any values
+TIMEOUT first(): Aborting a signal rejects the Promise with an AbortError DOMException Test timed out
```
Specifically this line is what's failing: https://github.com/WebKit/WebKit/blob/2d7df09cf3f08a332003e47dd0664f4c3423604a/LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-first.any.js#L66

cdumez commented on Source/WebCore/dom/InternalObserverFirst.cpp:
```diff
@@ -0,0 +1,111 @@
+
+    Ref observer = InternalObserverFirst::create(context, WTFMove(signal), WTFMove(promise));
+
+    observable.subscribeInternal(context, observer, SubscribeOptions { .signal = dependentSignal.ptr() });
```
`dependentSignal.ptr()` -> `WTFMove(dependentSignal)`

cdumez commented on Source/WebCore/dom/InternalObserverFirst.cpp:
```diff
@@ -0,0 +1,111 @@
+    void next(JSC::JSValue value) final
+    {
+        Ref { m_promise }->resolve<IDLAny>(value);
+        Ref { m_signal }->signalAbort(JSC::jsUndefined());
```
I wonder if this is really correct to pass `JSC::jsUndefined()` here, I suspect we may want `toJS(&globalObject, &globalObject, DOMException::create(ExceptionCode::AbortError)` instead. 

The spec says to signal abort controller. However, you opted NOT to use AbortController and use AbortSignal directly. Note how `AbortController::signalAbort()` is implemented:
```
void AbortController::abort(JSDOMGlobalObject& globalObject, JSC::JSValue reason)
{
    ASSERT(reason);
    if (reason.isUndefined())
        reason = toJS(&globalObject, &globalObject, DOMException::create(ExceptionCode::AbortError));

    protectedSignal()->signalAbort(reason);
}
```

Note the fallback for `reason`. This matches step 2 of https://dom.spec.whatwg.org/#abortsignal-signal-abort. 
I suspect our `reason` fallback should be in `AbortSignal::signalAbort()` instead of `AbortController::abort()` to match the spec exactly.

maraisr replied:
> I suspect our reason fallback should be in AbortSignal::signalAbort() instead of AbortController::abort() to match the spec exactly.

I can raise a bug/fix PR for this, and explore what that means aside from this change. For this PR specifically, I will use `AbortController` instead of AbortSignal. Think that way we aren't leaking another spec's detail here as well.

maraisr replied:
@cdumez I have raised a bug issue [here](https://bugs.webkit.org/show_bug.cgi?id=281961) with fix PR [here](https://github.com/WebKit/WebKit/pull/35618). I will raise another PR if this `.first` pr lands to remove the implementation detail once `signalAbort` does it.

cdumez commented on Source/WebCore/dom/InternalObserverFirst.cpp:
```diff
@@ -0,0 +1,114 @@
+
+    Vector<Ref<AbortSignal>> dependentSignals = { controller->protectedSignal() };
+    if (options.signal)
+        dependentSignals.append(*options.signal);
```
According to the safe C++ bot, should be:
```
dependentSignals.append(Ref { *options.signal });
```

cdumez commented on Source/WebCore/dom/InternalObserverFirst.cpp:
```diff
@@ -0,0 +1,114 @@
+private:
+    void next(JSC::JSValue value) final
+    {
+        auto* globalObject = scriptExecutionContext()->globalObject();
```
According to the safe C++ bot:
```
auto* globalObject = protectedScriptExecutionContext()->globalObject();
```

cdumez commented on Source/WebCore/dom/InternalObserverFirst.cpp:
```diff
@@ -0,0 +1,114 @@
+
+    Ref observer = InternalObserverFirst::create(context, WTFMove(controller), WTFMove(promise));
+
+    observable.subscribeInternal(context, observer, SubscribeOptions { .signal = WTFMove(dependentSignal) });
```
`WTFMove(observer)`

cdumez commented on Source/WebCore/dom/InternalObserverFirst.cpp:
```diff
@@ -0,0 +1,114 @@
+
+    Ref observer = InternalObserverFirst::create(context, WTFMove(controller), WTFMove(promise));
+
+    observable.subscribeInternal(context, observer, SubscribeOptions { .signal = WTFMove(dependentSignal) });
```
In terms of the test failure, I haven't had time to debug yet but I suspect the issue is that nobody is keeping `dependentSignal` alive, so it gets destroyed before it gets aborted.

You're passing it to `observable.subscribeInternal()`, `subscribeInternal()` will add an algorithm to `dependentSignal` but it will not store it anywhere. As a result, I'd expect it to get destroyed.

cdumez replied:
It appears I am right about the reason why the layout tests are failing. Nothing is keeping `dependentSignal` alive so it gets destroyed before we abort its child signal and thus it doesn't resolve the promise.

I have verified this is the cause by leaking the `dependentSignal`:
```
diff --git a/Source/WebCore/dom/InternalObserverFirst.cpp b/Source/WebCore/dom/InternalObserverFirst.cpp
index 1c83c3082a60..67b0e75b3314 100644
--- a/Source/WebCore/dom/InternalObserverFirst.cpp
+++ b/Source/WebCore/dom/InternalObserverFirst.cpp
@@ -98,6 +98,7 @@ void createInternalObserverOperatorFirst(ScriptExecutionContext& context, Observ
     if (options.signal)
         dependentSignals.append(*options.signal);
     Ref dependentSignal = AbortSignal::any(context, dependentSignals);
+    dependentSignal->ref(); // LEAK!
 
     if (dependentSignal->aborted())
         return promise->reject<IDLAny>(dependentSignal->reason().getValue());
```

So someone needs to keep the dependentSignal alive. Hopefully this helps you come up with a proper fix.

maraisr commented on Source/WebCore/dom/Observable.cpp:
```diff
@@ -70,16 +71,13 @@ void Observable::subscribe(ScriptExecutionContext& context, std::optional<Observ
-
-    if (options.signal)
-        subscriber->followSignal(*options.signal.get());
+    auto subscriber = Subscriber::create(context, observer, options);
```
The Subscriber should own/hold a reference to the options. `SubscribeOptions` is a simple struct, and does not implement `ActiveDOMObject` as such, we will pass by reference, and store the value inside the Subscriber.

Subscriber itself does implement `ActiveDOMObject`, and will be kept around as long as its callback does not fire `.complete`, or the subscriber itself goes out of scope (which `.last` is).

Seeing as promise-returning operators subscribe immediately in cpp land, we need to store the options on the Subscriber.

cc @cdumez

cdumez replied:
`auto` -> `Ref`

cdumez commented on LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-first.any.js:
```diff
@@ -72,7 +72,6 @@ promise_test(async () => {
       "Promise rejects with a DOMException for abortion");
   assert_equals(rejection.name, "AbortError",
       "Rejected with 'AbortError' DOMException");
-  assert_equals(rejection.message, "signal is aborted without reason");
```
Let's make sure to upstream this change to WPT. I agree enforcing a particular error message is not a good idea.

keithamus replied:
https://github.com/web-platform-tests/wpt/pull/48717/files WPTs have been altered upstream but this probably needs a sync or at least matching the changes properly.

maraisr replied:
I synced the file into this PR

cdumez commented on Source/WebCore/dom/Observable.cpp:
```diff
@@ -120,6 +118,11 @@ Ref<Observable> Observable::drop(ScriptExecutionContext& context, uint64_t amoun
     return create(createSubscriberCallbackDrop(context, *this, amount));
 }
 
+void Observable::first(ScriptExecutionContext& context, SubscribeOptions options, Ref<DeferredPromise>&& promise)
```
`SubscribeOptions` -> `const SubscribeOptions&`

cdumez commented on Source/WebCore/dom/Observable.h:
```diff
@@ -52,7 +52,7 @@ class Observable final : public ScriptWrappable, public RefCounted<Observable> {
 
     void subscribe(ScriptExecutionContext&, std::optional<ObserverUnion>, SubscribeOptions);
-    void subscribeInternal(ScriptExecutionContext&, Ref<InternalObserver>, SubscribeOptions);
+    void subscribeInternal(ScriptExecutionContext&, Ref<InternalObserver>, const SubscribeOptions&);
```
`Ref<InternalObserver>` -> `Ref<InternalObserver>&&`

cdumez commented on Source/WebCore/dom/Subscriber.cpp:
```diff
@@ -36,17 +36,20 @@
 namespace WebCore {
 
-Ref<Subscriber> Subscriber::create(ScriptExecutionContext& context, Ref<InternalObserver> observer)
+Ref<Subscriber> Subscriber::create(ScriptExecutionContext& context, Ref<InternalObserver> observer, const SubscribeOptions& options)
```
`Ref<InternalObserver>` -> `Ref<InternalObserver>&&`

cdumez commented on Source/WebCore/dom/Subscriber.cpp:
```diff
@@ -36,17 +36,20 @@
+Ref<Subscriber> Subscriber::create(ScriptExecutionContext& context, Ref<InternalObserver> observer, const SubscribeOptions& options)
 {
-    return adoptRef(*new Subscriber(context, observer));
+    return adoptRef(*new Subscriber(context, observer, options));
```
`WTFMove(observer)`

cdumez commented on Source/WebCore/dom/Subscriber.cpp:
```diff
@@ -36,17 +36,20 @@
     : ActiveDOMObject(&context)
     , m_abortController(AbortController::create(context))
     , m_observer(observer)
+    , m_options(options)
```
Yes, this seems fine.

cdumez commented on Source/WebCore/dom/Subscriber.cpp:
```diff
@@ -36,17 +36,20 @@
+    , m_options(options)
 {
     followSignal(m_abortController->signal());
+    if (options.signal)
```
```
if (RefPtr signal = options.signal)
    followSignal(*signal);
```
per our safe CPP guidelines.

cdumez commented on Source/WebCore/dom/Subscriber.h:
```diff
@@ -51,10 +52,9 @@ class Subscriber final : public ActiveDOMObject, public ScriptWrappable, public
     AbortSignal& signal() { return m_abortController->signal(); }
 
-    static Ref<Subscriber> create(ScriptExecutionContext&, Ref<InternalObserver>);
+    static Ref<Subscriber> create(ScriptExecutionContext&, Ref<InternalObserver>, const SubscribeOptions&);
```
`Ref<InternalObserver>` -> `Ref<InternalObserver>&&`

cdumez commented on Source/WebCore/dom/Subscriber.h:
```diff
@@ -51,10 +52,9 @@ class Subscriber final : public ActiveDOMObject, public ScriptWrappable, public
 
-    static Ref<Subscriber> create(ScriptExecutionContext&, Ref<InternalObserver>);
+    static Ref<Subscriber> create(ScriptExecutionContext&, Ref<InternalObserver>, const SubscribeOptions&);
+    explicit Subscriber(ScriptExecutionContext&, Ref<InternalObserver>, const SubscribeOptions&);
```
`Ref<InternalObserver>` -> `Ref<InternalObserver>&&`.

also, this constructor should be private (not public since there is a public `create()` function. I also don't think this needs `explicit` since there is more than 1 mandatory parameter.

cdumez commented on Source/WebCore/dom/Subscriber.h:
```diff
@@ -68,8 +68,11 @@ class Subscriber final : public ActiveDOMObject, public ScriptWrappable, public
     Lock m_teardownsLock;
     Ref<AbortController> m_abortController;
     Ref<InternalObserver> m_observer;
```
Data members should be the last things in the class as per our usual style. The private member functions should come first, then the private data members.

cdumez reviewed (approved):
A few minor comments but this looks correct.

maraisr committed:
Implement Observable#first
https://bugs.webkit.org/show_bug.cgi?id=277509

Reviewed by Chris Dumez and Darin Adler.

This change introduces the `.first` promise-returning operator
for Observables, using the InternalObserver and subscribing immediately.
This is required by the specification https://wicg.github.io/observable/#dom-observable-first

Because promise-returning operators subscribe immediatly in the runtime,
as opposed to user-land. We made a change to the Subscriber class to
now hold a reference to it's options, specifically any AbortSignals.

* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-last.any.js: Rebaselined.
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-first.any-expected.txt: Rebaselined.
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-first.any.worker-expected.txt: Rebaselined.
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-first.any.js:
  AbortErrors' description text is not normative,
   as such do not assert against it.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Observable.h:
* Source/WebCore/dom/Observable.cpp:
(WebCore::Observable::first): Added.
(WebCore::Observable::subscribeInternal):
  Move the abort signal following logic inside the Subscriber class.
* Source/WebCore/dom/Observable.idl:
  Exposes `Promise<any> first(options)` as a promise-returning operator.
* Source/WebCore/dom/Subscriber.h:
* Source/WebCore/dom/Subscriber.cpp:
(WebCore::Subscriber::create):
(WebCore::Subscriber::Subscriber):
  When a subscriber is created, it should own the options for the
  lifetime of the subscriber.
* Source/WebCore/dom/InternalObserverLast.cpp:
  Addressed diff feedback for `.first` in here as well.
* Source/WebCore/dom/InternalObserverLast.h:
* Source/WebCore/dom/InternalObserverFirst.cpp: Added.
(WebCore::createInternalObserverOperatorFirst):
* Source/WebCore/dom/InternalObserverFirst.h: Added.

Canonical link: https://commits.webkit.org/285650@main
